fixed regression in sparse of adjoint of sparse (#28948)

ExpandingMan opened the pull request:
This is a fix to #28948.  What was happening is that `sparse` was causing a call to be made to `SparseMatrixCSC` with explicit type parameters which was a fallback to an inefficient method, as without the methods I added only `SparseMatrixCSC` *without* explicit type parameters knew how to do this operation efficiently.

Note that there is probably room for improvement here: performance of `SparseMatrixCSC{T,Ti}(M::Adjoint{U,SparseMatrixCSC{U,Ti}})` where `T \ne U` will still be terrible, it's just not immediately obvious to me what the best solution to that is.

ExpandingMan committed:
fixed regression in sparse of adjoint

ExpandingMan commented:
Ok, what's the deal with these "xcode8" tests, do we expect these to past?  Looks like there are lots of failures that seem to have absolutely nothing to do with this PR, though looking around I see other PR's passing it.

vchuravy commented:
> Ok, what's the deal with these "xcode8" tests,
You mean the Mac OSX tests, yes we expect those to pass? There was one network failure so I restarted CI.

ExpandingMan commented:
For give my ignorance as to how the process works, but as this is essentially just a fix of a performance regression, I'm wondering if this can get a backport tag?

Sacha0 commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -399,6 +399,12 @@ function SparseMatrixCSC{Tv,Ti}(M::StridedMatrix) where {Tv,Ti}
+    copy(M)
+end
 SparseMatrixCSC(M::Adjoint{<:Any,<:SparseMatrixCSC}) = copy(M)
 SparseMatrixCSC(M::Transpose{<:Any,<:SparseMatrixCSC}) = copy(M)
```
Thanks @ExpandingMan! :)

Perhaps completing the method set would be worthwhile, e.g. the sketch
```
SparseMatrixCSC(M::Adjoint{<:Any,<:SparseMatrixCSC}) = copy(M)
SparseMatrixCSC(M::Transpose{<:Any,<:SparseMatrixCSC}) = copy(M)
SparseMatrixCSC{Tv}(M::Adjoint{Tv,<:SparseMatrixCSC{Tv}}) where {Tv} = copy(M)
SparseMatrixCSC{Tv}(M::Transpose{Tv,<:SparseMatrixCSC{Tv}}) where {Tv} = copy(M)
SparseMatrixCSC{Tv,Ti}(M::Adjoint{Tv,SparseMatrixCSC{Tv,Ti}}) where {Tv,Ti} = copy(M)
SparseMatrixCSC{Tv,Ti}(M::Transpose{Tv,SparseMatrixCSC{Tv,Ti}}) where {Tv,Ti} = copy(M)
```

Regarding

>Note that there is probably room for improvement here: performance of SparseMatrixCSC{T,Ti}(M::Adjoint{U,SparseMatrixCSC{U,Ti}}) where T \ne U will still be terrible, it's just not immediately obvious to me what the best solution to that is.

in combination with the preceding, the following sketch provides simple if suboptimal, but decent, implementations
```
SparseMatrixCSC{Tv}(M::Adjoint{<:Any,<:SparseMatrixCSC}) where {Tv} = SparseMatrixCSC{Tv}(copy(M))
SparseMatrixCSC{Tv}(M::Transpose{<:Any,<:SparseMatrixCSC}) where {Tv} = SparseMatrixCSC{Tv}(copy(M))
SparseMatrixCSC{Tv,Ti}(M::Adjoint{<:Any,<:SparseMatrixCSC}) where {Tv,Ti} = SparseMatrixCSC{Tv,Ti}(copy(M))
SparseMatrixCSC{Tv,Ti}(M::Transpose{<:Any,<:SparseMatrixCSC}) where {Tv,Ti} = SparseMatrixCSC{Tv,Ti}(copy(M))
```
The net number of method definitions could probably be reduced via `LinearAlgebra.AdjOrTrans`, e.g.
```
SparseMatrixCSC(M::Adjoint{<:Any,<:SparseMatrixCSC}) = copy(M)
SparseMatrixCSC(M::Transpose{<:Any,<:SparseMatrixCSC}) = copy(M)
```
collapses to
```
SparseMatrixCSC(M::AdjOrTrans{<:Any,<:SparseMatrixCSC}) = copy(M)
```
Best!

ExpandingMan committed:
added more methods

ExpandingMan commented:
Ok, added the recommended methods.  I did not use `LinearAlgebra.AdjOrTrans` since nothing else in `SparseArrays` seems to use that and I thought it would be confusing unless I changed everything.

Sacha0 commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -401,6 +401,24 @@ function SparseMatrixCSC{Tv,Ti}(M::StridedMatrix) where {Tv,Ti}
+end
+function SparseMatrixCSC{Tv,Ti}(M::Transpose{Tv,SparseMatrixCSC{Tv,Ti}}) where {Tv,Ti}
+    copy(M)
+end
```
Why use long-form definitions for these methods?

ExpandingMan committed:
collapsed unnecessary function blocks

ExpandingMan committed:
collapsed other function blocks

ExpandingMan commented:
I had left the other methods as blocks because they *definitely* go past my screen-split in vim :wink: .  I checked the file though, and there are already other lines which are longer, so I collapsed them as well.

Sacha0 reviewed (commented):
Looks great! Thanks @ExpandingMan! :)

Are these constructors already tested, or might they benefit from a little exercise?

ExpandingMan commented:
I have tested these, but not at all thoroughly.  If you want to do a quick sanity check it may be worthwhile.  The "really scary" regression definitely seems to be fixed.

(I think freebsd might need to be re-triggered.)

KristofferC commented:
It feels a bit "silly" that we need all these definitions but if that is what it takes let's just do it. Will merge when FreeBSD finishes CI. Butting a benchmark label so if someone feels inclined can put this case into BaseBenchmarks.jl

ExpandingMan commented:
One could probably do a thorough pass through `SparseArrays` and consolidate many things, but this was a very severe regression that needed to be fixed quickly and as I'm only somewhat familiar with the `SparseArrays` code I didn't want to be the one to start refactoring things.  Besides, this may be the sort of situation in which it pays to be explicit to the point of pedantry just to make sure issues like the one that prompted this PR don't crop up.

Is there some action required to re-trigger freebsd?

KristofferC commented:
Hm, I tried to do it manually but it didn't seem to take. Anyway, sparse tests succeeded on FreeBSD anyway so merging.
